File: crates/aiman_agent/src/benchmark.rs
use std::sync::Arc;
use std::time::Duration;

use anyhow::Context;
use chrono::Utc;
use serde::{Deserialize, Serialize};
use tokio::sync::Semaphore;

use aiman_shared::{EngineConfig, EngineInstance, EngineStatus, EngineType};

use crate::hardware::HardwareInfo;

#[derive(Debug, Deserialize)]
pub struct BenchmarkRequest {
    pub settings: BenchmarkRunSettings,
    pub host: Option<BenchmarkHostSnapshot>,
}

#[derive(Debug, Deserialize)]
pub struct BenchmarkRunSettings {
    pub concurrency: Option<Vec<usize>>,
    pub requests_per_concurrency: Option<usize>,
    pub prompt: Option<String>,
    pub prompt_words: Option<usize>,
    pub max_tokens: Option<u32>,
    pub temperature: Option<f32>,
    pub model: Option<String>,
    pub api_base_url: Option<String>,
    pub api_key: Option<String>,
    pub timeout_seconds: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BenchmarkSettings {
    pub concurrency: Vec<usize>,
    pub requests_per_concurrency: usize,
    pub prompt: String,
    pub prompt_words: usize,
    pub max_tokens: u32,
    pub temperature: f32,
    pub model: String,
    pub api_base_url: String,
    pub timeout_seconds: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BenchmarkHostSnapshot {
    pub id: String,
    pub name: String,
    pub base_url: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkRecord {
    pub id: String,
    pub ts: String,
    pub host: Option<BenchmarkHostSnapshot>,
    pub host_hardware: Option<HardwareInfo>,
    pub engine_config: EngineConfig,
    pub engine_status: EngineStatus,
    pub settings: BenchmarkSettings,
    pub results: Vec<BenchmarkResult>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkResult {
    pub concurrency: usize,
    pub requests: usize,
    pub success_count: usize,
    pub error_count: usize,
    pub duration_ms: u64,
    pub avg_latency_ms: u64,
    pub min_latency_ms: u64,
    pub max_latency_ms: u64,
    pub p50_latency_ms: u64,
    pub p90_latency_ms: u64,
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
    pub prompt_tps: f64,
    pub completion_tps: f64,
    pub requests_per_sec: f64,
    pub errors: Vec<String>,
}

#[derive(Debug)]
struct RequestOutcome {
    latency_ms: u64,
    prompt_tokens: u64,
    completion_tokens: u64,
    total_tokens: u64,
    error: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ModelListResponse {
    data: Vec<ModelInfo>,
}

#[derive(Debug, Deserialize)]
struct ModelInfo {
    id: String,
}

#[derive(Debug, Deserialize)]
struct ChatCompletionResponse {
    usage: Option<UsageInfo>,
}

#[derive(Debug, Deserialize)]
struct UsageInfo {
    prompt_tokens: Option<u64>,
    completion_tokens: Option<u64>,
    total_tokens: Option<u64>,
}

pub async fn run_benchmark(
    config: EngineConfig,
    instance: EngineInstance,
    host: Option<BenchmarkHostSnapshot>,
    host_hardware: Option<HardwareInfo>,
    request: BenchmarkRunSettings,
) -> anyhow::Result<BenchmarkRecord> {
    let defaults = NormalizedSettings::from_request(&config, request).await?;
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(defaults.timeout_seconds))
        .build()
        .context("create benchmark HTTP client")?;

    let mut results = Vec::new();
    for &concurrency in &defaults.concurrency {
        let outcome = run_concurrency(
            &client,
            &defaults,
            concurrency,
            defaults.requests_per_concurrency,
        )
        .await;
        results.push(outcome);
    }

    let record = BenchmarkRecord {
        id: format!("bench-{}", Utc::now().timestamp_millis()),
        ts: Utc::now().to_rfc3339(),
        host,
        host_hardware,
        engine_config: config,
        engine_status: instance.status,
        settings: BenchmarkSettings {
            concurrency: defaults.concurrency.clone(),
            requests_per_concurrency: defaults.requests_per_concurrency,
            prompt: defaults.prompt.clone(),
            prompt_words: defaults.prompt_words,
            max_tokens: defaults.max_tokens,
            temperature: defaults.temperature,
            model: defaults.model.clone(),
            api_base_url: defaults.api_base_url.clone(),
            timeout_seconds: defaults.timeout_seconds,
        },
        results,
    };

    Ok(record)
}

async fn run_concurrency(
    client: &reqwest::Client,
    settings: &NormalizedSettings,
    concurrency: usize,
    total_requests: usize,
) -> BenchmarkResult {
    let total_requests = total_requests.max(1);
    let concurrency = concurrency.max(1);
    let semaphore = Arc::new(Semaphore::new(concurrency));
    let start = tokio::time::Instant::now();
    let mut handles = Vec::with_capacity(total_requests);

    for _ in 0..total_requests {
        let semaphore = semaphore.clone();
        let client = client.clone();
        let settings = settings.clone();
        handles.push(tokio::spawn(async move {
            let permit = match semaphore.acquire_owned().await {
                Ok(permit) => permit,
                Err(_) => {
                    return RequestOutcome {
                        latency_ms: 0,
                        prompt_tokens: 0,
                        completion_tokens: 0,
                        total_tokens: 0,
                        error: Some("semaphore closed".to_string()),
                    }
                }
            };
            let _permit = permit;
            run_request(&client, &settings).await
        }));
    }

    let mut outcomes = Vec::with_capacity(total_requests);
    for handle in handles {
        match handle.await {
            Ok(outcome) => outcomes.push(outcome),
            Err(err) => outcomes.push(RequestOutcome {
                latency_ms: 0,
                prompt_tokens: 0,
                completion_tokens: 0,
                total_tokens: 0,
                error: Some(format!("worker join error: {err}")),
            }),
        }
    }

    let duration = start.elapsed();
    summarize_results(concurrency, total_requests, duration, outcomes)
}

async fn run_request(
    client: &reqwest::Client,
    settings: &NormalizedSettings,
) -> RequestOutcome {
    let start = tokio::time::Instant::now();
    let req = client
        .post(format!("{}/v1/chat/completions", settings.api_base_url))
        .json(&serde_json::json!({
            "model": &settings.model,
            "messages": [
                { "role": "user", "content": &settings.prompt }
            ],
            "max_tokens": settings.max_tokens,
            "temperature": settings.temperature
        }));

    let req = if let Some(api_key) = settings.api_key.as_deref() {
        req.bearer_auth(api_key)
    } else {
        req
    };

    let response = match req.send().await {
        Ok(resp) => resp,
        Err(err) => {
            return RequestOutcome {
                latency_ms: start.elapsed().as_millis() as u64,
                prompt_tokens: 0,
                completion_tokens: 0,
                total_tokens: 0,
                error: Some(format!("request failed: {err}")),
            }
        }
    };

    let status = response.status();
    if !status.is_success() {
        let body = response.text().await.unwrap_or_default();
        return RequestOutcome {
            latency_ms: start.elapsed().as_millis() as u64,
            prompt_tokens: 0,
            completion_tokens: 0,
            total_tokens: 0,
            error: Some(format!("HTTP {status}: {body}")),
        };
    }

    let body = match response.json::<ChatCompletionResponse>().await {
        Ok(body) => body,
        Err(err) => {
            return RequestOutcome {
                latency_ms: start.elapsed().as_millis() as u64,
                prompt_tokens: 0,
                completion_tokens: 0,
                total_tokens: 0,
                error: Some(format!("decode failed: {err}")),
            }
        }
    };

    let usage = body.usage.unwrap_or(UsageInfo {
        prompt_tokens: Some(0),
        completion_tokens: Some(0),
        total_tokens: Some(0),
    });

    let prompt_tokens = usage.prompt_tokens.unwrap_or(0);
    let completion_tokens = usage.completion_tokens.unwrap_or(0);
    let total_tokens = usage
        .total_tokens
        .unwrap_or(prompt_tokens + completion_tokens);

    RequestOutcome {
        latency_ms: start.elapsed().as_millis() as u64,
        prompt_tokens,
        completion_tokens,
        total_tokens,
        error: None,
    }
}

fn summarize_results(
    concurrency: usize,
    requests: usize,
    duration: Duration,
    outcomes: Vec<RequestOutcome>,
) -> BenchmarkResult {
    let duration_ms = duration.as_millis() as u64;
    let duration_secs = duration.as_secs_f64().max(0.001);
    let mut latencies = Vec::new();
    let mut prompt_tokens = 0u64;
    let mut completion_tokens = 0u64;
    let mut total_tokens = 0u64;
    let mut errors = Vec::new();

    for outcome in outcomes.iter() {
        if let Some(err) = outcome.error.as_ref() {
            if errors.len() < 6 {
                errors.push(err.clone());
            }
            continue;
        }
        latencies.push(outcome.latency_ms);
        prompt_tokens += outcome.prompt_tokens;
        completion_tokens += outcome.completion_tokens;
        total_tokens += outcome.total_tokens;
    }

    let success_count = latencies.len();
    let error_count = requests.saturating_sub(success_count);
    latencies.sort_unstable();
    let avg_latency_ms = if success_count > 0 {
        latencies.iter().sum::<u64>() / success_count as u64
    } else {
        0
    };
    let min_latency_ms = latencies.first().copied().unwrap_or(0);
    let max_latency_ms = latencies.last().copied().unwrap_or(0);
    let p50_latency_ms = percentile(&latencies, 0.5);
    let p90_latency_ms = percentile(&latencies, 0.9);

    BenchmarkResult {
        concurrency,
        requests,
        success_count,
        error_count,
        duration_ms,
        avg_latency_ms,
        min_latency_ms,
        max_latency_ms,
        p50_latency_ms,
        p90_latency_ms,
        prompt_tokens,
        completion_tokens,
        total_tokens,
        prompt_tps: prompt_tokens as f64 / duration_secs,
        completion_tps: completion_tokens as f64 / duration_secs,
        requests_per_sec: success_count as f64 / duration_secs,
        errors,
    }
}

fn percentile(values: &[u64], pct: f64) -> u64 {
    if values.is_empty() {
        return 0;
    }
    let rank = ((values.len() as f64 - 1.0) * pct).round() as usize;
    values[rank.min(values.len() - 1)]
}

#[derive(Clone)]
struct NormalizedSettings {
    concurrency: Vec<usize>,
    requests_per_concurrency: usize,
    prompt: String,
    prompt_words: usize,
    max_tokens: u32,
    temperature: f32,
    model: String,
    api_base_url: String,
    api_key: Option<String>,
    timeout_seconds: u64,
}

impl NormalizedSettings {
    async fn from_request(
        config: &EngineConfig,
        request: BenchmarkRunSettings,
    ) -> anyhow::Result<Self> {
        let concurrency = request
            .concurrency
            .unwrap_or_else(|| vec![1, 2, 4, 8])
            .into_iter()
            .filter(|value| *value > 0)
            .collect::<Vec<_>>();
        let concurrency = if concurrency.is_empty() {
            vec![1]
        } else {
            concurrency
        };
        let requests_per_concurrency = request.requests_per_concurrency.unwrap_or(8).max(1);
        let requested_words = request.prompt_words.unwrap_or(120).max(1);
        let prompt = request
            .prompt
            .filter(|value| !value.trim().is_empty())
            .unwrap_or_else(|| generate_prompt(requested_words));
        let prompt_words = count_words(&prompt).max(1);
        let max_tokens = request.max_tokens.unwrap_or(256).max(1);
        let temperature = request.temperature.unwrap_or(0.2).clamp(0.0, 2.0);
        let api_base_url = request
            .api_base_url
            .and_then(|value| normalize_base_url(&value))
            .or_else(|| infer_api_base(config))
            .context("unable to infer engine API base URL")?;

        let model = match request.model {
            Some(model) if !model.trim().is_empty() => model,
            _ => fetch_default_model(&api_base_url, request.api_key.as_deref()).await?,
        };

        Ok(Self {
            concurrency,
            requests_per_concurrency,
            prompt,
            prompt_words,
            max_tokens,
            temperature,
            model,
            api_base_url,
            api_key: request.api_key,
            timeout_seconds: request.timeout_seconds.unwrap_or(90).max(10),
        })
    }
}

async fn fetch_default_model(api_base_url: &str, api_key: Option<&str>) -> anyhow::Result<String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(20))
        .build()
        .context("create model discovery client")?;
    let req = client.get(format!("{}/v1/models", api_base_url));
    let req = if let Some(api_key) = api_key {
        req.bearer_auth(api_key)
    } else {
        req
    };
    let response = req.send().await.context("fetch model list")?;
    let status = response.status();
    if !status.is_success() {
        let body = response.text().await.unwrap_or_default();
        anyhow::bail!("model list request failed (HTTP {}): {}", status, body);
    }
    let body = response.json::<ModelListResponse>().await?;
    body.data
        .first()
        .map(|model| model.id.clone())
        .context("model list returned no models")
}

fn normalize_base_url(value: &str) -> Option<String> {
    let trimmed = value.trim().trim_end_matches('/');
    if trimmed.is_empty() {
        return None;
    }
    Some(trimmed.to_string())
}

fn infer_api_base(config: &EngineConfig) -> Option<String> {
    let mut host = parse_arg_value(&config.args, "--host")
        .or_else(|| parse_arg_value(&config.args, "--bind"))
        .unwrap_or_else(|| "127.0.0.1".to_string());
    let port = parse_arg_value(&config.args, "--port")
        .and_then(|value| value.parse::<u16>().ok())
        .unwrap_or_else(|| default_port(&config.engine_type));
    if host == "0.0.0.0" || host == "::" {
        host = "127.0.0.1".to_string();
    }
    Some(format!("http://{host}:{port}"))
}

fn parse_arg_value(args: &[String], key: &str) -> Option<String> {
    for (idx, value) in args.iter().enumerate() {
        if value == key {
            return args.get(idx + 1).cloned();
        }
        if let Some(stripped) = value.strip_prefix(&format!("{key}=")) {
            return Some(stripped.to_string());
        }
    }
    None
}

fn default_port(engine_type: &EngineType) -> u16 {
    match engine_type {
        // Llama.cpp-derived servers typically default to 8080.
        EngineType::LlamaCpp | EngineType::IkLlamaCpp => 8080,
        // FastLLM uses 8080 in its default server example.
        EngineType::Fastllm => 8080,
        // vLLM + most others default to 8000.
        _ => 8000,
    }
}

fn generate_prompt(words: usize) -> String {
    const WORDS: &[&str] = &[
        "ocean", "signal", "ember", "circuit", "memory", "harbor", "silent", "gravity", "silver",
        "atlas", "garden", "vector", "timber", "echo", "planet", "canvas", "mirror", "thread",
        "story", "nebula", "glacier", "pixel", "horizon", "compass", "lattice", "whisper", "orchid",
        "shadow", "river", "lantern",
    ];
    let mut prompt = String::new();
    for idx in 0..words {
        if idx > 0 {
            prompt.push(' ');
        }
        prompt.push_str(WORDS[idx % WORDS.len()]);
    }
    prompt
}

fn count_words(prompt: &str) -> usize {
    prompt.split_whitespace().filter(|word| !word.is_empty()).count()
}

File: dashboard/src/server/benchmark.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import type {
  BenchmarkRecord,
  BenchmarkResult,
  DockerImage,
  EngineConfig,
  EngineInstance,
  HardwareInfo,
  HostConfig,
  NormalizedBenchmarkSettings,
  RequestOutcome
} from "./types";

export type { BenchmarkRecord, BenchmarkResult };

// Path is injected at module init to keep this module testable.
let dashboardBenchmarksPath: string;

export function initBenchmarkPath(opts: { dashboardBenchmarksPath: string }) {
  dashboardBenchmarksPath = opts.dashboardBenchmarksPath;
}

export async function appendDashboardBenchmark(record: BenchmarkRecord) {
  const dir = path.dirname(dashboardBenchmarksPath);
  await mkdir(dir, { recursive: true });
  const line = `${JSON.stringify(record)}\n`;
  await writeFile(dashboardBenchmarksPath, line, { flag: "a" });
}

export async function readDashboardBenchmarks(): Promise<BenchmarkRecord[]> {
  try {
    const raw = await readFile(dashboardBenchmarksPath, "utf8");
    if (!raw.trim()) {
      return [];
    }
    const records: BenchmarkRecord[] = [];
    for (const line of raw.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }
      try {
        records.push(JSON.parse(trimmed) as BenchmarkRecord);
      } catch {
        continue;
      }
    }
    return records;
  } catch {
    return [];
  }
}

export async function runDashboardBenchmark(
  host: HostConfig,
  engineId: string,
  payload: Record<string, unknown>
): Promise<BenchmarkRecord> {
  const settingsPayload =
    "settings" in payload && typeof payload.settings === "object"
      ? (payload.settings as Record<string, unknown>)
      : payload;

  const [config, instance, hardware] = await Promise.all([
    fetchConfig(host, engineId),
    fetchInstance(host, engineId),
    fetchHardware(host)
  ]);

  const image =
    config.engine_type === "Docker" && config.docker?.image_id
      ? await fetchImage(host, config.docker.image_id)
      : null;

  if (instance.status !== "Running") {
    throw new Error(`engine is not running (status ${instance.status})`);
  }

  const resolved = await normalizeBenchmarkSettings(config, host, settingsPayload, image ?? undefined);
  const results: BenchmarkResult[] = [];
  for (const concurrency of resolved.concurrency) {
    const result = await runBenchmarkConcurrency(resolved, concurrency);
    results.push(result);
  }

  return {
    id: `bench-${Date.now()}`,
    ts: new Date().toISOString(),
    origin: "dashboard",
    host: { id: host.id, name: host.name, base_url: host.base_url },
    host_hardware: hardware,
    engine_config: config,
    engine_status: instance.status,
    settings: {
      concurrency: resolved.concurrency,
      requests_per_concurrency: resolved.requestsPerConcurrency,
      prompt: resolved.prompt,
      prompt_words: resolved.promptWords,
      max_tokens: resolved.maxTokens,
      temperature: resolved.temperature,
      model: resolved.model,
      api_base_url: resolved.apiBaseUrl,
      timeout_seconds: resolved.timeoutSeconds
    },
    results
  };
}

async function fetchConfig(host: HostConfig, engineId: string): Promise<EngineConfig> {
  const res = await fetch(`${host.base_url}/v1/configs`, {
    headers: host.api_key ? { Authorization: `Bearer ${host.api_key}` } : undefined
  });
  if (!res.ok) {
    throw new Error(`failed to load configs (HTTP ${res.status})`);
  }
  const body = (await res.json()) as { configs?: EngineConfig[] };
  const config = (body.configs ?? []).find((item) => item.id === engineId);
  if (!config) {
    throw new Error("engine config not found");
  }
  return config;
}

async function fetchImage(host: HostConfig, imageId: string): Promise<DockerImage | null> {
  const res = await fetch(`${host.base_url}/v1/images/${encodeURIComponent(imageId)}`, {
    headers: host.api_key ? { Authorization: `Bearer ${host.api_key}` } : undefined
  });
  if (!res.ok) {
    return null;
  }
  const body = (await res.json()) as { image?: DockerImage };
  return body.image ?? null;
}

async function fetchInstance(host: HostConfig, engineId: string): Promise<EngineInstance> {
  const res = await fetch(`${host.base_url}/v1/engines/${engineId}`, {
    headers: host.api_key ? { Authorization: `Bearer ${host.api_key}` } : undefined
  });
  if (!res.ok) {
    throw new Error(`failed to load engine (HTTP ${res.status})`);
  }
  const body = (await res.json()) as { instance: EngineInstance };
  return body.instance;
}

async function fetchHardware(host: HostConfig): Promise<HardwareInfo | null> {
  const res = await fetch(`${host.base_url}/v1/hardware`, {
    headers: host.api_key ? { Authorization: `Bearer ${host.api_key}` } : undefined
  });
  if (!res.ok) {
    return null;
  }
  const body = (await res.json()) as { hardware?: HardwareInfo };
  return body.hardware ?? null;
}

async function normalizeBenchmarkSettings(
  config: EngineConfig,
  host: HostConfig,
  payload: Record<string, unknown>,
  image?: DockerImage
): Promise<NormalizedBenchmarkSettings> {
  const concurrency = parseConcurrency(payload.concurrency).filter((value) => value > 0);
  const resolvedConcurrency = concurrency.length ? concurrency : [1, 2, 4, 8];
  const requestsPerConcurrency = clampNumber(payload.requests_per_concurrency, 8, 1);
  const promptWords = clampNumber(payload.prompt_words, 120, 1);
  const prompt =
    typeof payload.prompt === "string" && payload.prompt.trim().length
      ? payload.prompt.trim()
      : generatePrompt(promptWords);
  const promptWordCount = countWords(prompt);
  const maxTokens = clampNumber(payload.max_tokens, 256, 1);
  const temperature = clampNumber(payload.temperature, 0.2, 0);
  const apiBaseUrl =
    typeof payload.api_base_url === "string" && payload.api_base_url.trim().length
      ? normalizeBaseUrl(payload.api_base_url)
      : inferApiBase(config, host, image);
  if (!apiBaseUrl) {
    throw new Error("unable to infer engine API base URL");
  }

  const apiKey =
    typeof payload.api_key === "string" && payload.api_key.trim().length
      ? payload.api_key.trim()
      : undefined;
  const model =
    typeof payload.model === "string" && payload.model.trim().length
      ? payload.model.trim()
      : await fetchDefaultModel(apiBaseUrl, apiKey);
  const timeoutSeconds = clampNumber(payload.timeout_seconds, 90, 10);

  return {
    concurrency: resolvedConcurrency,
    requestsPerConcurrency,
    prompt,
    promptWords: promptWordCount,
    maxTokens,
    temperature,
    model,
    apiBaseUrl,
    apiKey,
    timeoutSeconds
  };
}

async function runBenchmarkConcurrency(
  settings: NormalizedBenchmarkSettings,
  concurrency: number
): Promise<BenchmarkResult> {
  const totalRequests = Math.max(1, settings.requestsPerConcurrency);
  const limiter = createLimiter(concurrency);
  const start = Date.now();
  const tasks = Array.from({ length: totalRequests }, () =>
    limiter(() => runBenchmarkRequest(settings))
  );
  const outcomes = await Promise.all(tasks);
  const durationMs = Math.max(1, Date.now() - start);
  const latencies: number[] = [];
  let promptTokens = 0;
  let completionTokens = 0;
  let totalTokens = 0;
  const errors: string[] = [];

  outcomes.forEach((outcome) => {
    if (outcome.error) {
      if (errors.length < 6) {
        errors.push(outcome.error);
      }
      return;
    }
    latencies.push(outcome.latency_ms);
    promptTokens += outcome.prompt_tokens;
    completionTokens += outcome.completion_tokens;
    totalTokens += outcome.total_tokens;
  });

  latencies.sort((a, b) => a - b);
  const successCount = latencies.length;
  const durationSecs = durationMs / 1000;
  const avgLatency =
    successCount > 0 ? Math.round(latencies.reduce((sum, value) => sum + value, 0) / successCount) : 0;

  return {
    concurrency,
    requests: totalRequests,
    success_count: successCount,
    error_count: totalRequests - successCount,
    duration_ms: durationMs,
    avg_latency_ms: avgLatency,
    min_latency_ms: latencies[0] ?? 0,
    max_latency_ms: latencies[latencies.length - 1] ?? 0,
    p50_latency_ms: percentile(latencies, 0.5),
    p90_latency_ms: percentile(latencies, 0.9),
    prompt_tokens: promptTokens,
    completion_tokens: completionTokens,
    total_tokens: totalTokens,
    prompt_tps: durationSecs > 0 ? promptTokens / durationSecs : 0,
    completion_tps: durationSecs > 0 ? completionTokens / durationSecs : 0,
    requests_per_sec: durationSecs > 0 ? successCount / durationSecs : 0,
    errors
  };
}

async function runBenchmarkRequest(settings: NormalizedBenchmarkSettings): Promise<RequestOutcome> {
  const start = Date.now();
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), settings.timeoutSeconds * 1000);
  try {
    const res = await fetch(`${settings.apiBaseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...(settings.apiKey ? { Authorization: `Bearer ${settings.apiKey}` } : {})
      },
      body: JSON.stringify({
        model: settings.model,
        messages: [{ role: "user", content: settings.prompt }],
        max_tokens: settings.maxTokens,
        temperature: settings.temperature
      }),
      signal: controller.signal
    });

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      return {
        latency_ms: Date.now() - start,
        prompt_tokens: 0,
        completion_tokens: 0,
        total_tokens: 0,
        error: `HTTP ${res.status}: ${text}`
      };
    }

    const body = (await res.json()) as {
      usage?: { prompt_tokens?: number; completion_tokens?: number; total_tokens?: number };
    };
    const usage = body.usage ?? {};
    const promptTokens = usage.prompt_tokens ?? 0;
    const completionTokens = usage.completion_tokens ?? 0;
    const totalTokens = usage.total_tokens ?? promptTokens + completionTokens;

    return {
      latency_ms: Date.now() - start,
      prompt_tokens: promptTokens,
      completion_tokens: completionTokens,
      total_tokens: totalTokens
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : "request failed";
    return {
      latency_ms: Date.now() - start,
      prompt_tokens: 0,
      completion_tokens: 0,
      total_tokens: 0,
      error: message
    };
  } finally {
    clearTimeout(timeout);
  }
}

function createLimiter(limit: number) {
  let active = 0;
  const queue: Array<() => void> = [];
  const runNext = () => {
    if (active >= limit) {
      return;
    }
    const next = queue.shift();
    if (!next) {
      return;
    }
    active += 1;
    next();
  };

  return <T>(task: () => Promise<T>): Promise<T> =>
    new Promise((resolve, reject) => {
      const run = () => {
        task()
          .then(resolve, reject)
          .finally(() => {
            active = Math.max(0, active - 1);
            runNext();
          });
      };
      queue.push(run);
      runNext();
    });
}

function percentile(values: number[], pct: number) {
  if (!values.length) {
    return 0;
  }
  const index = Math.min(values.length - 1, Math.round((values.length - 1) * pct));
  return values[index];
}

function clampNumber(value: unknown, fallback: number, minValue?: number) {
  const parsed = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(parsed)) {
    return fallback;
  }
  if (minValue !== undefined && parsed < minValue) {
    return minValue;
  }
  return parsed;
}

function parseConcurrency(value: unknown): number[] {
  if (Array.isArray(value)) {
    return value.map((item) => Number(item)).filter((item) => Number.isFinite(item));
  }
  if (typeof value === "string") {
    return value
      .split(",")
      .map((item) => Number(item.trim()))
      .filter((item) => Number.isFinite(item));
  }
  return [];
}

function normalizeBaseUrl(value: string) {
  const trimmed = value.trim().replace(/\/$/, "");
  return trimmed;
}

function inferApiBase(config: EngineConfig, host: HostConfig, image?: DockerImage) {
  const dockerArgs = config.docker?.args ?? [];
  const dockerPorts = [
    ...(image?.ports ?? []),
    ...(config.docker?.extra_ports ?? [])
  ];
  const hostValue =
    parseArgValue(dockerArgs, "--host") ??
    parseArgValue(dockerArgs, "--bind") ??
    parseArgValue(config.args, "--host") ??
    parseArgValue(config.args, "--bind") ??
    "127.0.0.1";
  const portValue =
    parseArgValue(dockerArgs, "--port") ??
    parseArgValue(config.args, "--port") ??
    parseDockerHostPort(dockerPorts);
  const port = portValue ? Number(portValue) : defaultPort(config.engine_type);
  if (!Number.isFinite(port)) {
    return null;
  }
  const url = new URL(host.base_url);
  const isLocal =
    hostValue === "127.0.0.1" ||
    hostValue === "0.0.0.0" ||
    hostValue === "::" ||
    hostValue === "localhost";
  const resolvedHost = isLocal ? url.hostname : hostValue;
  return `${url.protocol}//${resolvedHost}:${port}`;
}

function parseArgValue(args: string[], key: string) {
  for (let index = 0; index < args.length; index += 1) {
    const value = args[index];
    if (value === key) {
      return args[index + 1];
    }
    if (value.startsWith(`${key}=`)) {
      return value.slice(key.length + 1);
    }
  }
  return null;
}

function parseDockerHostPort(ports: string[]) {
  for (const mapping of ports) {
    const port = parseDockerPortMapping(mapping);
    if (port) {
      return String(port);
    }
  }
  return null;
}

function parseDockerPortMapping(mapping: string) {
  const trimmed = mapping.trim();
  if (!trimmed) {
    return null;
  }
  const noProto = trimmed.split("/")[0] ?? trimmed;
  const parts = noProto.split(":");
  const hostPort = parts.length >= 2 ? parts[parts.length - 2] : parts[0];
  const parsed = Number(hostPort);
  if (!Number.isFinite(parsed)) {
    return null;
  }
  return parsed;
}

function defaultPort(engineType: EngineConfig["engine_type"]) {
  // Defaults mirror engine vendors to make benchmark inference sensible.
  if (engineType === "LlamaCpp" || engineType === "ik_llamacpp") {
    return 8080;
  }
  if (engineType === "fastllm") {
    return 8080;
  }
  return 8000;
}

function generatePrompt(words: number) {
  const pool = [
    "ocean",
    "signal",
    "ember",
    "circuit",
    "memory",
    "harbor",
    "silent",
    "gravity",
    "silver",
    "atlas",
    "garden",
    "vector",
    "timber",
    "echo",
    "planet",
    "canvas",
    "mirror",
    "thread",
    "story",
    "nebula",
    "glacier",
    "pixel",
    "horizon",
    "compass",
    "lattice",
    "whisper",
    "orchid",
    "shadow",
    "river",
    "lantern"
  ];
  const parts = Array.from({ length: words }, (_, idx) => pool[idx % pool.length]);
  return parts.join(" ");
}

function countWords(value: string) {
  return value.split(/\s+/).filter(Boolean).length;
}

async function fetchDefaultModel(apiBaseUrl: string, apiKey?: string) {
  const res = await fetch(`${apiBaseUrl}/v1/models`, {
    headers: apiKey ? { Authorization: `Bearer ${apiKey}` } : undefined
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`model list request failed (HTTP ${res.status}): ${text}`);
  }
  const body = (await res.json()) as { data?: Array<{ id: string }> };
  const model = body.data?.[0]?.id;
  if (!model) {
    throw new Error("model list returned no models");
  }
  return model;
}
